File: backend_rust/src/repository.rs
use chrono::Utc;
use sqlx::PgPool;
use uuid::Uuid;

use crate::models::Quote;

/// Репозиторий для работы с цитатами
pub struct QuoteRepository {
    pool: PgPool,
}

impl QuoteRepository {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    /// Возвращает случайную цитату
    pub async fn get_random(&self) -> anyhow::Result<Quote> {
        use tracing::warn;
        
        // Сначала проверяем, есть ли цитаты вообще
        let count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM quotes")
            .fetch_one(&self.pool)
            .await?;
        
        if count == 0 {
            warn!("No quotes found in database");
            return Err(anyhow::anyhow!("no quotes found"));
        }
        
        let quote = sqlx::query_as::<_, Quote>(
            "SELECT id, text, author, likes_count, created_at, updated_at 
             FROM quotes 
             ORDER BY RANDOM() 
             LIMIT 1"
        )
        .fetch_optional(&self.pool)
        .await?;

        quote.ok_or_else(|| {
            warn!("Query returned no rows despite count being {}", count);
            anyhow::anyhow!("no quotes found")
        })
    }

    /// Возвращает все цитаты с пагинацией и поиском
    pub async fn get_all(
        &self,
        page: i32,
        page_size: i32,
        search: Option<&str>,
    ) -> anyhow::Result<(Vec<Quote>, i64)> {
        let offset = (page - 1) * page_size;

        // Подсчет общего количества
        let total = if let Some(search_term) = search {
            let search_pattern = format!("%{}%", search_term);
            sqlx::query_scalar::<_, i64>(
                "SELECT COUNT(*) FROM quotes WHERE text ILIKE $1 OR author ILIKE $1"
            )
            .bind(&search_pattern)
            .fetch_one(&self.pool)
            .await?
        } else {
            sqlx::query_scalar::<_, i64>("SELECT COUNT(*) FROM quotes")
                .fetch_one(&self.pool)
                .await?
        };

        // Получение цитат
        let quotes = if let Some(search_term) = search {
            let search_pattern = format!("%{}%", search_term);
            sqlx::query_as::<_, Quote>(
                "SELECT id, text, author, likes_count, created_at, updated_at 
                 FROM quotes 
                 WHERE text ILIKE $1 OR author ILIKE $1 
                 ORDER BY created_at DESC 
                 LIMIT $2 OFFSET $3"
            )
            .bind(&search_pattern)
            .bind(page_size)
            .bind(offset)
            .fetch_all(&self.pool)
            .await?
        } else {
            sqlx::query_as::<_, Quote>(
                "SELECT id, text, author, likes_count, created_at, updated_at 
                 FROM quotes 
                 ORDER BY created_at DESC 
                 LIMIT $1 OFFSET $2"
            )
            .bind(page_size)
            .bind(offset)
            .fetch_all(&self.pool)
            .await?
        };

        Ok((quotes, total))
    }

    /// Возвращает цитату по ID
    pub async fn get_by_id(&self, id: &str) -> anyhow::Result<Quote> {
        let quote = sqlx::query_as::<_, Quote>(
            "SELECT id, text, author, likes_count, created_at, updated_at 
             FROM quotes 
             WHERE id = $1"
        )
        .bind(id)
        .fetch_optional(&self.pool)
        .await?;

        quote.ok_or_else(|| anyhow::anyhow!("quote not found"))
    }

    /// Создает новую цитату
    pub async fn create(&self, quote: &Quote) -> anyhow::Result<()> {
        sqlx::query(
            "INSERT INTO quotes (id, text, author, likes_count, created_at, updated_at)
             VALUES ($1, $2, $3, $4, $5, $6)"
        )
        .bind(&quote.id)
        .bind(&quote.text)
        .bind(&quote.author)
        .bind(quote.likes_count)
        .bind(quote.created_at)
        .bind(quote.updated_at)
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    /// Обновляет существующую цитату
    pub async fn update(&self, id: &str, quote: &Quote) -> anyhow::Result<()> {
        let result = sqlx::query(
            "UPDATE quotes 
             SET text = $1, author = $2, updated_at = $3
             WHERE id = $4"
        )
        .bind(&quote.text)
        .bind(&quote.author)
        .bind(Utc::now())
        .bind(id)
        .execute(&self.pool)
        .await?;

        if result.rows_affected() == 0 {
            return Err(anyhow::anyhow!("quote not found"));
        }

        Ok(())
    }

    /// Удаляет цитату
    pub async fn delete(&self, id: &str) -> anyhow::Result<()> {
        let result = sqlx::query("DELETE FROM quotes WHERE id = $1")
            .bind(id)
            .execute(&self.pool)
            .await?;

        if result.rows_affected() == 0 {
            return Err(anyhow::anyhow!("quote not found"));
        }

        Ok(())
    }

    /// Ставит лайк цитате
    pub async fn like(&self, id: &str, user_ip: &str, user_agent: Option<&str>) -> anyhow::Result<()> {
        // Начинаем транзакцию
        let mut tx = self.pool.begin().await?;

        // Проверяем, не лайкал ли уже этот пользователь эту цитату
        let existing_like: Option<String> = sqlx::query_scalar(
            "SELECT id FROM likes 
             WHERE quote_id = $1 AND user_ip = $2
             FOR UPDATE"
        )
        .bind(id)
        .bind(user_ip)
        .fetch_optional(&mut *tx)
        .await?;

        if existing_like.is_some() {
            return Err(anyhow::anyhow!("you have already liked this quote"));
        }

        // Увеличиваем счетчик лайков
        let result = sqlx::query(
            "UPDATE quotes 
             SET likes_count = likes_count + 1, updated_at = $1
             WHERE id = $2"
        )
        .bind(Utc::now())
        .bind(id)
        .execute(&mut *tx)
        .await?;

        if result.rows_affected() == 0 {
            return Err(anyhow::anyhow!("quote not found"));
        }

        // Сохраняем информацию о лайке
        let like_id = Uuid::new_v4().to_string();
        sqlx::query(
            "INSERT INTO likes (id, quote_id, user_ip, user_agent, created_at)
             VALUES ($1, $2, $3, $4, $5)
             ON CONFLICT (quote_id, user_ip) DO NOTHING"
        )
        .bind(&like_id)
        .bind(id)
        .bind(user_ip)
        .bind(user_agent)
        .bind(Utc::now())
        .execute(&mut *tx)
        .await?;

        tx.commit().await?;
        Ok(())
    }

    /// Проверяет, лайкнул ли пользователь цитату
    pub async fn is_liked(&self, quote_id: &str, user_ip: &str) -> anyhow::Result<bool> {
        let count: i64 = sqlx::query_scalar(
            "SELECT COUNT(*) FROM likes WHERE quote_id = $1 AND user_ip = $2"
        )
        .bind(quote_id)
        .bind(user_ip)
        .fetch_one(&self.pool)
        .await?;

        Ok(count > 0)
    }

    /// Возвращает топ цитату за неделю
    pub async fn get_top_weekly(&self) -> anyhow::Result<Quote> {
        let quote = sqlx::query_as::<_, Quote>(
            "SELECT id, text, author, likes_count, created_at, updated_at 
             FROM quotes 
             WHERE created_at >= NOW() - INTERVAL '7 days'
             ORDER BY likes_count DESC, created_at DESC
             LIMIT 1"
        )
        .fetch_optional(&self.pool)
        .await?;

        quote.ok_or_else(|| anyhow::anyhow!("no quotes found for the last week"))
    }

    /// Возвращает топ цитату за всё время
    pub async fn get_top_all_time(&self) -> anyhow::Result<Quote> {
        let quote = sqlx::query_as::<_, Quote>(
            "SELECT id, text, author, likes_count, created_at, updated_at 
             FROM quotes 
             ORDER BY likes_count DESC, created_at DESC
             LIMIT 1"
        )
        .fetch_optional(&self.pool)
        .await?;

        quote.ok_or_else(|| anyhow::anyhow!("no quotes found"))
    }

    /// Сбрасывает все лайки
    pub async fn reset_likes(&self) -> anyhow::Result<()> {
        let mut tx = self.pool.begin().await?;

        // Обнуляем счетчики лайков
        sqlx::query("UPDATE quotes SET likes_count = 0, updated_at = $1")
            .bind(Utc::now())
            .execute(&mut *tx)
            .await?;

        // Удаляем все записи о лайках
        sqlx::query("DELETE FROM likes")
            .execute(&mut *tx)
            .await?;

        tx.commit().await?;
        Ok(())
    }
}

File: backend_cpp/src/repository.cpp
#include "repository.h"
#include <libpq-fe.h>
#include <stdexcept>
#include <sstream>
#include <algorithm>

Quote QuoteRepository::rowToQuote(PGresult* res, int row) {
    Quote quote;
    quote.id = PQgetvalue(res, row, 0);
    quote.text = PQgetvalue(res, row, 1);
    quote.author = PQgetvalue(res, row, 2);
    quote.likesCount = std::stoi(PQgetvalue(res, row, 3));
    quote.createdAt = PQgetvalue(res, row, 4);
    quote.updatedAt = PQgetvalue(res, row, 5);
    return quote;
}

Quote QuoteRepository::getRandom() {
    const char* query = "SELECT id, text, author, likes_count, created_at, updated_at FROM quotes ORDER BY RANDOM() LIMIT 1";
    PGresult* res = PQexec(db_->getConnection(), query);
    
    if (PQresultStatus(res) != PGRES_TUPLES_OK || PQntuples(res) == 0) {
        PQclear(res);
        throw std::runtime_error("no quotes found");
    }
    
    Quote quote = rowToQuote(res, 0);
    PQclear(res);
    return quote;
}

std::pair<std::vector<Quote>, int64_t> QuoteRepository::getAll(int page, int pageSize, const std::string& search) {
    int offset = (page - 1) * pageSize;
    std::vector<Quote> quotes;
    int64_t total = 0;
    
    PGconn* conn = db_->getConnection();
    
    // Count total
    if (!search.empty()) {
        std::string searchPattern = "%" + search + "%";
        const char* countQuery = "SELECT COUNT(*) FROM quotes WHERE text ILIKE $1 OR author ILIKE $1";
        const char* paramValues[1] = {searchPattern.c_str()};
        int paramLengths[1] = {static_cast<int>(search.length())};
        int paramFormats[1] = {0};
        
        int paramLengths[1] = {static_cast<int>(searchPattern.length())};
        PGresult* countRes = PQexecParams(conn, countQuery, 1, nullptr, paramValues, paramLengths, paramFormats, 0);
        if (PQresultStatus(countRes) == PGRES_TUPLES_OK) {
            total = std::stoll(PQgetvalue(countRes, 0, 0));
        }
        PQclear(countRes);
        
        // Get quotes with search
        const char* query = "SELECT id, text, author, likes_count, created_at, updated_at FROM quotes WHERE text ILIKE $1 OR author ILIKE $1 ORDER BY created_at DESC LIMIT $2 OFFSET $3";
        const char* values[3] = {searchPattern.c_str(), std::to_string(pageSize).c_str(), std::to_string(offset).c_str()};
        int lengths[3] = {static_cast<int>(searchPattern.length()), -1, -1};
        int formats[3] = {0, 0, 0};
        
        PGresult* res = PQexecParams(conn, query, 3, nullptr, values, lengths, formats, 0);
        if (PQresultStatus(res) == PGRES_TUPLES_OK) {
            int ntuples = PQntuples(res);
            quotes.reserve(ntuples);
            for (int i = 0; i < ntuples; ++i) {
                quotes.push_back(rowToQuote(res, i));
            }
        }
        PQclear(res);
    } else {
        const char* countQuery = "SELECT COUNT(*) FROM quotes";
        PGresult* countRes = PQexec(conn, countQuery);
        if (PQresultStatus(countRes) == PGRES_TUPLES_OK) {
            total = std::stoll(PQgetvalue(countRes, 0, 0));
        }
        PQclear(countRes);
        
        // Get quotes
        std::string query = "SELECT id, text, author, likes_count, created_at, updated_at FROM quotes ORDER BY created_at DESC LIMIT " + 
                           std::to_string(pageSize) + " OFFSET " + std::to_string(offset);
        PGresult* res = PQexec(conn, query.c_str());
        if (PQresultStatus(res) == PGRES_TUPLES_OK) {
            int ntuples = PQntuples(res);
            quotes.reserve(ntuples);
            for (int i = 0; i < ntuples; ++i) {
                quotes.push_back(rowToQuote(res, i));
            }
        }
        PQclear(res);
    }
    
    return {quotes, total};
}

Quote QuoteRepository::getById(const std::string& id) {
    const char* query = "SELECT id, text, author, likes_count, created_at, updated_at FROM quotes WHERE id = $1";
    const char* values[1] = {id.c_str()};
    int lengths[1] = {static_cast<int>(id.length())};
    int formats[1] = {0};
    
    PGresult* res = PQexecParams(db_->getConnection(), query, 1, nullptr, values, lengths, formats, 0);
    
    if (PQresultStatus(res) != PGRES_TUPLES_OK || PQntuples(res) == 0) {
        PQclear(res);
        throw std::runtime_error("quote not found");
    }
    
    Quote quote = rowToQuote(res, 0);
    PQclear(res);
    return quote;
}

void QuoteRepository::create(const Quote& quote) {
    const char* query = "INSERT INTO quotes (id, text, author, likes_count, created_at, updated_at) VALUES ($1, $2, $3, $4, $5, $6)";
    const char* values[6] = {
        quote.id.c_str(),
        quote.text.c_str(),
        quote.author.c_str(),
        std::to_string(quote.likesCount).c_str(),
        quote.createdAt.c_str(),
        quote.updatedAt.c_str()
    };
    int lengths[6] = {
        static_cast<int>(quote.id.length()),
        static_cast<int>(quote.text.length()),
        static_cast<int>(quote.author.length()),
        -1, -1, -1
    };
    int formats[6] = {0, 0, 0, 0, 0, 0};
    
    PGresult* res = PQexecParams(db_->getConnection(), query, 6, nullptr, values, lengths, formats, 0);
    if (PQresultStatus(res) != PGRES_COMMAND_OK) {
        std::string error = PQerrorMessage(db_->getConnection());
        PQclear(res);
        throw std::runtime_error("failed to create quote: " + error);
    }
    PQclear(res);
}

void QuoteRepository::update(const std::string& id, const Quote& quote) {
    const char* query = "UPDATE quotes SET text = $1, author = $2, updated_at = NOW() WHERE id = $3";
    const char* values[3] = {quote.text.c_str(), quote.author.c_str(), id.c_str()};
    int lengths[3] = {
        static_cast<int>(quote.text.length()),
        static_cast<int>(quote.author.length()),
        static_cast<int>(id.length())
    };
    int formats[3] = {0, 0, 0};
    
    PGresult* res = PQexecParams(db_->getConnection(), query, 3, nullptr, values, lengths, formats, 0);
    if (PQresultStatus(res) != PGRES_COMMAND_OK || PQcmdTuples(res)[0] == '0') {
        PQclear(res);
        throw std::runtime_error("quote not found");
    }
    PQclear(res);
}

void QuoteRepository::deleteQuote(const std::string& id) {
    const char* query = "DELETE FROM quotes WHERE id = $1";
    const char* values[1] = {id.c_str()};
    int lengths[1] = {static_cast<int>(id.length())};
    int formats[1] = {0};
    
    PGresult* res = PQexecParams(db_->getConnection(), query, 1, nullptr, values, lengths, formats, 0);
    if (PQresultStatus(res) != PGRES_COMMAND_OK || PQcmdTuples(res)[0] == '0') {
        PQclear(res);
        throw std::runtime_error("quote not found");
    }
    PQclear(res);
}

void QuoteRepository::like(const std::string& id, const std::string& userIP, const std::string& userAgent) {
    PGconn* conn = db_->getConnection();
    
    // Begin transaction
    PGresult* beginRes = PQexec(conn, "BEGIN");
    PQclear(beginRes);
    
    try {
        // Check if already liked
        const char* checkQuery = "SELECT id FROM likes WHERE quote_id = $1 AND user_ip = $2 FOR UPDATE";
        const char* checkValues[2] = {id.c_str(), userIP.c_str()};
        int checkLengths[2] = {static_cast<int>(id.length()), static_cast<int>(userIP.length())};
        int checkFormats[2] = {0, 0};
        
        PGresult* checkRes = PQexecParams(conn, checkQuery, 2, nullptr, checkValues, checkLengths, checkFormats, 0);
        if (PQntuples(checkRes) > 0) {
            PQclear(checkRes);
            PQexec(conn, "ROLLBACK");
            throw std::runtime_error("you have already liked this quote");
        }
        PQclear(checkRes);
        
        // Update likes count
        const char* updateQuery = "UPDATE quotes SET likes_count = likes_count + 1, updated_at = NOW() WHERE id = $1";
        const char* updateValues[1] = {id.c_str()};
        int updateLengths[1] = {static_cast<int>(id.length())};
        int updateFormats[1] = {0};
        
        PGresult* updateRes = PQexecParams(conn, updateQuery, 1, nullptr, updateValues, updateLengths, updateFormats, 0);
        if (PQresultStatus(updateRes) != PGRES_COMMAND_OK || PQcmdTuples(updateRes)[0] == '0') {
            PQclear(updateRes);
            PQexec(conn, "ROLLBACK");
            throw std::runtime_error("quote not found");
        }
        PQclear(updateRes);
        
        // Insert like (using gen_random_uuid() for UUID generation)
        const char* insertQuery = "INSERT INTO likes (id, quote_id, user_ip, user_agent, created_at) VALUES (gen_random_uuid()::text, $1, $2, $3, NOW()) ON CONFLICT (quote_id, user_ip) DO NOTHING";
        const char* insertValues[3] = {id.c_str(), userIP.c_str(), userAgent.c_str()};
        int insertLengths[3] = {
            static_cast<int>(id.length()),
            static_cast<int>(userIP.length()),
            static_cast<int>(userAgent.length())
        };
        int insertFormats[3] = {0, 0, 0};
        
        PGresult* insertRes = PQexecParams(conn, insertQuery, 3, nullptr, insertValues, insertLengths, insertFormats, 0);
        PQclear(insertRes);
        
        // Commit
        PGresult* commitRes = PQexec(conn, "COMMIT");
        PQclear(commitRes);
    } catch (...) {
        PQexec(conn, "ROLLBACK");
        throw;
    }
}

bool QuoteRepository::isLiked(const std::string& quoteId, const std::string& userIP) {
    const char* query = "SELECT COUNT(*) FROM likes WHERE quote_id = $1 AND user_ip = $2";
    const char* values[2] = {quoteId.c_str(), userIP.c_str()};
    int lengths[2] = {
        static_cast<int>(quoteId.length()),
        static_cast<int>(userIP.length())
    };
    int formats[2] = {0, 0};
    
    PGresult* res = PQexecParams(db_->getConnection(), query, 2, nullptr, values, lengths, formats, 0);
    bool result = false;
    if (PQresultStatus(res) == PGRES_TUPLES_OK) {
        result = std::stoi(PQgetvalue(res, 0, 0)) > 0;
    }
    PQclear(res);
    return result;
}

std::unordered_map<std::string, bool> QuoteRepository::areLiked(const std::vector<std::string>& quoteIds, const std::string& userIP) {
    std::unordered_map<std::string, bool> result;
    
    if (quoteIds.empty()) {
        return result;
    }
    
    // Build query with IN clause (simpler and works reliably)
    std::string query = "SELECT quote_id FROM likes WHERE quote_id IN (";
    std::vector<const char*> values(quoteIds.size() + 1);
    std::vector<int> lengths(quoteIds.size() + 1);
    std::vector<int> formats(quoteIds.size() + 1, 0);
    
    for (size_t i = 0; i < quoteIds.size(); ++i) {
        if (i > 0) query += ", ";
        query += "$" + std::to_string(i + 1);
        values[i] = quoteIds[i].c_str();
        lengths[i] = static_cast<int>(quoteIds[i].length());
        result[quoteIds[i]] = false; // Initialize to false
    }
    query += ") AND user_ip = $" + std::to_string(quoteIds.size() + 1);
    values[quoteIds.size()] = userIP.c_str();
    lengths[quoteIds.size()] = static_cast<int>(userIP.length());
    
    PGresult* res = PQexecParams(db_->getConnection(), query.c_str(), 
                                 static_cast<int>(values.size()), nullptr, 
                                 values.data(), lengths.data(), formats.data(), 0);
    
    if (PQresultStatus(res) == PGRES_TUPLES_OK) {
        int ntuples = PQntuples(res);
        for (int i = 0; i < ntuples; ++i) {
            std::string quoteId = PQgetvalue(res, i, 0);
            result[quoteId] = true;
        }
    }
    PQclear(res);
    
    return result;
}

Quote QuoteRepository::getTopWeekly() {
    const char* query = "SELECT id, text, author, likes_count, created_at, updated_at FROM quotes WHERE created_at >= NOW() - INTERVAL '7 days' ORDER BY likes_count DESC, created_at DESC LIMIT 1";
    PGresult* res = PQexec(db_->getConnection(), query);
    
    if (PQresultStatus(res) != PGRES_TUPLES_OK || PQntuples(res) == 0) {
        PQclear(res);
        throw std::runtime_error("no quotes found for the last week");
    }
    
    Quote quote = rowToQuote(res, 0);
    PQclear(res);
    return quote;
}

Quote QuoteRepository::getTopAllTime() {
    const char* query = "SELECT id, text, author, likes_count, created_at, updated_at FROM quotes ORDER BY likes_count DESC, created_at DESC LIMIT 1";
    PGresult* res = PQexec(db_->getConnection(), query);
    
    if (PQresultStatus(res) != PGRES_TUPLES_OK || PQntuples(res) == 0) {
        PQclear(res);
        throw std::runtime_error("no quotes found");
    }
    
    Quote quote = rowToQuote(res, 0);
    PQclear(res);
    return quote;
}

void QuoteRepository::resetLikes() {
    PGconn* conn = db_->getConnection();
    
    PGresult* beginRes = PQexec(conn, "BEGIN");
    PQclear(beginRes);
    
    try {
        PGresult* updateRes = PQexec(conn, "UPDATE quotes SET likes_count = 0, updated_at = NOW()");
        if (PQresultStatus(updateRes) != PGRES_COMMAND_OK) {
            PQclear(updateRes);
            PQexec(conn, "ROLLBACK");
            throw std::runtime_error("failed to reset likes");
        }
        PQclear(updateRes);
        
        PGresult* deleteRes = PQexec(conn, "DELETE FROM likes");
        PQclear(deleteRes);
        
        PGresult* commitRes = PQexec(conn, "COMMIT");
        PQclear(commitRes);
    } catch (...) {
        PQexec(conn, "ROLLBACK");
        throw;
    }
}

